src/download_px4_logfile/mavctrl/ftp.go

```
@@ -0,0 +1,549 @@
+					} else {
+						item.Size = i
+					}
+					item_data_buf = ""
+					// Item ready, add to list
+					list = append(list, item)
+					parse_state = 0
+				} else {
+					// fill the size buffer
+					item_data_buf += string(data[i])
+				}
+			}
+		}
+	}
+	return
+}
+
+func (ctrl *MavlinkControl) DownloadLogFile(src_dir string, src_file string, dst_path string, fsize int) bool {
+	// Download file
+	var src_file_path string = "/fs/microsd/log/" + src_dir + "/" + src_file
+	var dst_file_path string = dst_path + "__part"
+
+	if !ctrl.ftpOpenFileRO(src_file_path) {
+		log.Printf("[ftp] Open file '%s' failed!\n", src_file_path)
+		return false
+	}
+	log.Printf("[ftp] Reading...")
+
+	var offset uint32 = 0
+	//var readsize uint32 = ftpReadBufferSize
+	buf := []uint8{}
+	file, err := os.Create(dst_file_path)
+	defer file.Close()
+	if err != nil {
+		log.Println("[ftp] ERROR to create file: ", err.Error())
+		return false
+	}
+
+	for {
+		slice, size := ctrl.ftpReadFile(offset)
+		if size == 0 {
+			break
+		}
+		offset += uint32(size)
+		/*		err := ioutil.WriteFile(dst_file_path, slice, 0644)
+				if err != nil {
+					log.Printf("[ftp] File '%s' write failed: %d", dst_file_path, err)
+					return false
+				}
+		*/
+		buf = append(buf, slice...)
+		if len(buf) > ftpReadBufferSize {
+			log.Printf("[ftp] Progress %d%%", (offset*100)/uint32(fsize))
+			_, err := file.Write(buf)
+			//err := ioutil.WriteFile(dst_file_path, buf, 0644)
+			if err != nil {
+				log.Printf("[ftp] File '%s' write failed: %d", dst_file_path, err)
+				return false
+			}
+			buf = []uint8{}
+		}
+	}
+	log.Printf("[ftp] Progress 100%%\n")
+	log.Printf("[ftp] Reading done (%d bytes)\n", offset)
+	ctrl.ftpTerminateSession()
+
+	if len(buf) > 0 {
+		_, err := file.Write(buf)
+		if err != nil {
+			log.Printf("[ftp] File '%s' write failed: %d", dst_file_path, err)
+			return false
+		}
+	}
+	file.Close()
+
+	os.Rename(dst_file_path, dst_path)
```

This can fail, the error is not checked